File: NetworkEmulator/maze.py
import random
import pickle
# import pygame
from enum import Enum

class Cell():
    def __init__(self, x=-1, y=-1):
        self.x = x
        self.y = y
        self.visited = False
        self.north = True
        self.east = True
        self.south = True
        self.west = True
        self.final = False
    def __repr__(self):
        return "%d,%d\t%s N:%s E:%s S:%s W:%s"%(self.x, self.y, str(self.visited),
            str(self.north), str(self.east), str(self.south), str(self.west))

"""Maze class

Contains the database of the mgit@github.com:Jojojoppe/CPS_EPROJ.gitaze and generator functions
"""
class Maze():
    def __init__(self, w=16, h=16):
        self.width = w
        self.height = h
        self.initial = (0, 0)

        self.grid = {}
        for h in range(self.height):
            for w in range(self.width):
                self.grid[(w,h)] = Cell(w, h)

    def generate(self):
        stack = []
        self.grid[self.initial].visited = True
        stack.append(self.initial)
        while len(stack)>0:
            current = stack.pop()
            neighbours = self.get_neighbours(self.grid[current])
            if len(neighbours)>0:
                stack.append(current)
                r = random.randint(0, len(neighbours)-1)
                chosen = neighbours[r]
                # Remove wall between
                if self.grid[current].x == self.grid[chosen].x:
                    if self.grid[current].y > self.grid[chosen].y:
                        self.grid[current].north = False
                        self.grid[chosen].south = False
                    else:
                        self.grid[current].south = False
                        self.grid[chosen].north = False
                elif self.grid[current].y == self.grid[chosen].y:
                    if self.grid[current].x > self.grid[chosen].x:
                        self.grid[current].west = False
                        self.grid[chosen].east = False
                    else:
                        self.grid[current].east = False
                        self.grid[chosen].west = False
                self.grid[chosen].visited = True
                stack.append(chosen)
        # Generate random endpoint
        side = random.randint(0,3)
        # North
        if side==0:
            nr = random.randint(0,self.width-1)
            self.grid[(nr, 0)].north = False
            self.grid[(nr, 0)].final = True
        # South
        elif side==1:
            nr = random.randint(0,self.width-1)
            self.grid[(nr, self.height-1)].south = False
            self.grid[(nr, self.height-1)].final = True
        # East
        elif side==2:
            nr = random.randint(0,self.height-1)
            self.grid[(self.width-1, nr)].east = False
            self.grid[(self.width-1, nr)].final = True
        # West
        elif side==3:
            nr = random.randint(0,self.height-1)
            self.grid[(0, nr)].west = False
            self.grid[(0, nr)].final = True

    def get_neighbours(self, c:Cell):
        res = []
        if (c.x+1,c.y) in self.grid and not self.grid[(c.x+1,c.y)].visited: res.append((c.x+1, c.y))
        if (c.x-1,c.y) in self.grid and not self.grid[(c.x-1,c.y)].visited: res.append((c.x-1, c.y))
        if (c.x,c.y+1) in self.grid and not self.grid[(c.x,c.y+1)].visited: res.append((c.x, c.y+1))
        if (c.x,c.y-1) in self.grid and not self.grid[(c.x,c.y-1)].visited: res.append((c.x, c.y-1))
        return res

    def print_grid(self):
        for h in range(self.height):
            for w in range(self.width):
                print(self.grid[(w,h)])
            
    @classmethod
    def load(cls, path):
        f = open(path, "rb")
        maze = pickle.loads(f.read())
        f.close()
        return maze

    def save(self, path):
        f = open(path, "wb")
        f.write(pickle.dumps(self))
        f.close()

    def dumps(self):
        return pickle.dumps(self)

    @classmethod
    def loads(cls, data:bytes):
        return pickle.loads(data)

    """Get info of position
    Returns tupe with:
        north, east, south, west: True if wall, False if open
        final: True if exit of maze
    """
    def getInfo(self, pos):
        c = self.grid[pos]
        return c.north, c.east, c.south, c.west, c.final

File: NetEmuC/src/Maze.cpp
#include "Maze.hpp"
#include <cstdlib>
#include <tuple>
#include <time.h>

Maze::Maze(int width, int height, std::tuple<int,int> exit){
    this->width = width;
    this->height = height;
    this->exit = exit;
    
    // Fill the grid
    for(int h=0; h<height; h++){
        for(int w=0; w<width; w++){
            Cell c = {w, h, false, true, true, true, true, false};
            this->grid[std::tuple<int,int>(w,h)] = c;
        }
    }
    this->data = NULL;
}

Maze::~Maze(){
    if(this->data==NULL){
        return;
    }
    free(this->data);
}

void Maze::generate(){
     srand(time(0)); 
    std::vector<std::tuple<int,int>> stack;
    std::tuple<int,int> initial(0, 0);

    this->grid[initial].visited = true;
    stack.push_back(initial);
    while(stack.size()>0){
        std::tuple<int,int> current = stack.back();
        stack.pop_back();
        std::vector<std::tuple<int,int>> neighbours = this->get_neighbours(current);
        if(neighbours.size()>0){
            stack.push_back(current);
            int r = rand()%neighbours.size();
            std::tuple<int,int> chosen = neighbours[r];
            // Remove wall between
            if(this->grid[current].x == this->grid[chosen].x){
                if(this->grid[current].y > this->grid[chosen].y){
                    this->grid[current].north = false;
                    this->grid[chosen].south = false;
                }else{
                    this->grid[current].south = false;
                    this->grid[chosen].north = false;
                }
            }else if(this->grid[current].y == this->grid[chosen].y){
                if(this->grid[current].x > this->grid[chosen].x){
                    this->grid[current].west = false;
                    this->grid[chosen].east = false;
                }else{
                    this->grid[current].east = false;
                    this->grid[chosen].west = false;
                }
            }
            this->grid[chosen].visited = true;
            stack.push_back(chosen);
        }
    }

    // Exit
    this->grid[this->exit].final = true;
    printf("Place exit on %d,%d\r\n", std::get<0>(this->exit), std::get<1>(this->exit));

    // Fill maze data
    this->dataLen = this->width*this->height*13;
    this->data = (uint8_t*) malloc(this->dataLen);
    for(int x=0; x<this->width; x++){
        for(int y=0; y<this->height; y++){
            ((uint32_t*)(this->data+13*this->height*x+y*13))[0] = (uint32_t)x;
            ((uint32_t*)(this->data+13*this->height*x+y*13))[1] = (uint32_t)y;
            uint8_t* bs = (uint8_t*) &((uint32_t*)(this->data+13*this->height*x+y*13))[2];
            bs[0] = (uint8_t) this->grid[std::tuple<int,int>(x,y)].north;
            bs[1] = (uint8_t) this->grid[std::tuple<int,int>(x,y)].east;
            bs[2] = (uint8_t) this->grid[std::tuple<int,int>(x,y)].south;
            bs[3] = (uint8_t) this->grid[std::tuple<int,int>(x,y)].west;
            bs[4] = (uint8_t) this->grid[std::tuple<int,int>(x,y)].final;
        }
    }
}

std::vector<std::tuple<int,int>> Maze::get_neighbours(std::tuple<int,int> pos){
    std::vector<std::tuple<int,int>> res;
    int x = std::get<0>(pos);
    int y = std::get<1>(pos);
    if(x+1>-1&&x+1<this->width && y>-1&&y<this->height && !this->grid[std::tuple<int,int>(x+1,y)].visited){
        res.push_back(std::tuple<int,int>(x+1,y));
    }
    if(x-1>-1&&x-1<this->width && y>-1&&y<this->height && !this->grid[std::tuple<int,int>(x-1,y)].visited){
        res.push_back(std::tuple<int,int>(x-1,y));
    }
    if(x>-1&&x<this->width && y+1>-1&&y+1<this->height && !this->grid[std::tuple<int,int>(x,y+1)].visited){
        res.push_back(std::tuple<int,int>(x,y+1));
    }
    if(x>-1&&x<this->width && y-1>-1&&y-1<this->height && !this->grid[std::tuple<int,int>(x,y-1)].visited){
        res.push_back(std::tuple<int,int>(x,y-1));
    }
    return res;
}

void Maze::reload(){
    for(int x=0; x<this->width; x++){
        for(int y=0; y<this->height; y++){
            uint8_t* bs = (uint8_t*) &((uint32_t*)(this->data+13*this->height*x+y*13))[2];
            this->grid[std::tuple<int,int>(x,y)].north = (bool)bs[0];
            this->grid[std::tuple<int,int>(x,y)].east = (bool)bs[1];
            this->grid[std::tuple<int,int>(x,y)].south = (bool)bs[2];
            this->grid[std::tuple<int,int>(x,y)].west = (bool)bs[3];
            this->grid[std::tuple<int,int>(x,y)].final = (bool)bs[4];
        }
    }
}